tensorflow/compiler/xla/service/gpu/alpa_nccl_wrapper.cc

```diff
@@ -0,0 +1,234 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+// This file implements nccl apis for alpa to use. 
+
+#include "tensorflow/compiler/xla/service/gpu/alpa_nccl_wrapper.h"
+
+#include <memory>
+#include <utility>
+#include <stdexcept>
+
+#include "absl/container/flat_hash_map.h"
+#include "absl/strings/str_format.h"
+#include "absl/synchronization/mutex.h"
+#include "absl/time/time.h"
+#include "tensorflow/compiler/xla/debug_options_flags.h"
+#include "tensorflow/compiler/xla/service/global_device_id.h"
+#include "tensorflow/compiler/xla/service/gpu/gpu_executable_run_options.h"
+#include "tensorflow/compiler/xla/status_macros.h"
+#include "tensorflow/compiler/xla/statusor.h"
+#include "tensorflow/core/platform/env.h"
+
+#include "pybind11/pybind11.h"
+#include "pybind11/attr.h"
+#include "pybind11/cast.h"
+#include "pybind11/numpy.h"
+#include "pybind11/pytypes.h"
+#include "pybind11/stl_bind.h"
+
+#include "tensorflow/compiler/xla/python/py_buffer.h"
+#include "tensorflow/compiler/xla/python/py_client.h"
+
+PYBIND11_MAKE_OPAQUE(std::vector<ncclComm_t>);
+
+namespace xla {
+namespace gpu {
+
+ncclDataType_t ToNcclDataType(PrimitiveType element_type) {
+  switch (element_type) {
+    case S8:
+      return ncclInt8;
+    case PRED:
+    case U8:
+      return ncclUint8;
+    case S32:
+      return ncclInt32;
+    case U32:
+      return ncclUint32;
+    case S64:
+      return ncclInt64;
+    case U64:
+      return ncclUint64;
+    case F16:
+      return ncclFloat16;
+    case F32:
+    case C64:
+      return ncclFloat32;
+    case F64:
+    case C128:
+      return ncclFloat64;
+  }
+}
+
+int SizeOfType(ncclDataType_t element_type) {
+  switch (element_type) {
+    case ncclInt8:
+      return 1;
+    case ncclUint8:
+      return 1;
+    case ncclInt32:
+      return 4;
+    case ncclUint32:
+      return 4;
+    case ncclInt64:
+      return 8;
+    case ncclUint64:
+      return 8;
+    case ncclFloat16:
+      return 2;
+    case ncclFloat32:
+      return 4;
+    case ncclFloat64:
+      return 8;
+    default:
+      return 4;
+  }
+}
+
+std::shared_ptr< std::vector<ncclComm_t> > NcclInitCommunicator(std::vector<int> devices_vec) {
+    int n_devices = devices_vec.size();
+    std::vector<ncclComm_t> comms;
+    comms.resize(n_devices);
+    ncclCommInitAll(comms.data(), n_devices, devices_vec.data());
+    return std::make_shared< std::vector<ncclComm_t> >(std::move(comms));
+}
+
+void NcclLocalAllGather(std::vector<ncclComm_t> comms, 
+                         std::vector<PyBuffer::object> buffers, 
```

Comment:
A style issue:
```suggestion
void NcclLocalAllGather(std::vector<ncclComm_t> comms, 
                        std::vector<PyBuffer::object> buffers, 
```
Also apply the changes to other functions.
Or you can use 
`clang-format -i tensorflow/compiler/xla/service/gpu/alpa_nccl_wrapper.cc` to auto-format your c++ code

Reply:
OK. Finished.